config/initializers/spree.rb

```diff
@@ -8,6 +8,16 @@
 
 require 'spree/product_filters'
 
+# Due to a bug in ActiveRecord we need to load the tagging code in Gateway which
+# should have inherited it from its parent PaymentMethod.
+# We have to call it before loading the PaymentMethod decorator because the
+# tagging code won't load twice within the inheritance chain.
+# https://github.com/openfoodfoundation/openfoodnetwork/issues/3121
+Spree::Gateway.class_eval do
+  acts_as_taggable
+  attr_accessible :tag_list
+end
```

Comment:
but if essentially this is a order-related bug doesn't a `require 'spree/payment_method'` in the appropriate place helps us here? apart from that, from https://github.com/rails/rails/issues/3847 looks like some gem could be messing things up as well.

Reply:
Yes, it is order-related. Yes, it could be a gem. It could even be Spree.

A `require 'spree/payment_method'` doesn't help, because the inheritance already happened when we access the Spree code. I tried that in many places, for hours, after Luis tried that in many places, for hours. You can have another go and maybe you can find a better solution. But we spent several dev days on this already and couldn't find any better solution.

Since this bug is hopefully fixed in the next version of Rails we shouldn't invest much more time.

One experiment I would support as a tech-debt task: Instead of adding tags to `Spree::PaymentMethod`, we could try to have a new class called `TaggablePaymentMethod` which basically decorates any other payment method. All code using tagging interface would use that class instead of using the payment method instance directly.

What do you think? Should I create an issue for that?

Reply:
It's fine. It doesn't make sense to spend any more time on this. I wouldn't add the tech debt task either for now.